File: LogisimPlus/LogisimPlus/logisim/data/Direction.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;
using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.data
{
	public sealed class Direction : AttributeOptionInterface
	{

		public static readonly Direction East = new Direction("East", InnerEnum.East, 0);
		public static readonly Direction North = new Direction("North", InnerEnum.North, 1);
		public static readonly Direction West = new Direction("West", InnerEnum.West, 2);
		public static readonly Direction South = new Direction("South", InnerEnum.South, 3);

		private static readonly List<Direction> valueList = new List<Direction>();

		static Direction()
		{
			valueList.Add(East);
			valueList.Add(North);
			valueList.Add(West);
			valueList.Add(South);
		}

		public enum InnerEnum
		{
			East,
			North,
			West,
			South
		}

		public readonly InnerEnum innerEnumValue;
		private readonly string nameValue;
		private readonly int ordinalValue;
		private static int nextOrdinal = 0;

		private int id;

		private Direction(string name, InnerEnum innerEnum, int id)
		{
			this.id = id;

			nameValue = name;
			ordinalValue = nextOrdinal++;
			innerEnumValue = innerEnum;
		}

		/// @deprecated Please use valueOf instead 
		[Obsolete("Please use valueOf instead")]
		public static Direction parse(string str)
		{
			string cap = str.Substring(0, 1).ToUpper() + str.Substring(1);
			return Direction.valueOf(cap);
		}

		public static readonly Direction[] cardinals = new Direction[] {East, North, West, South};

		public static Direction[] Cardinals
		{
			get
			{
				return cardinals;
			}
		}

		public static Direction fromInt(int x)
		{
			return cardinals[x];
		}

		private static string[] displayStrings = new string[] {Strings.get("directionEastOption"), Strings.get("directionNorthOption"), Strings.get("directionWestOption"), Strings.get("directionSouthOption")};

		public string toDisplayString()
		{
			return displayStrings[id];
		}

		private static string[] displayStringsVertical = new string[] {Strings.get("directionEastVertical"), Strings.get("directionNorthVertical"), Strings.get("directionWestVertical"), Strings.get("directionSouthVertical")};

		public string toVerticalDisplayString()
		{
			return displayStringsVertical[id];
		}

		public double toRadians()
		{
			return id * Math.PI / 2.0;
		}

		public int toDegrees()
		{
			return id * 90;
		}

		public Direction reverse()
		{
			return fromInt((id + 2) & 3);
		}

		public Direction rotateCW()
		{
			return fromInt((id - 1) & 3);
		}

		public Direction rotateCCW()
		{
			return fromInt((id + 1) & 3);
		}

		// for AttributeOptionInterface
		public object Value
		{
			get
			{
				return this;
			}
		}

		public static Direction[] values()
		{
			return valueList.ToArray();
		}

		public int ordinal()
		{
			return ordinalValue;
		}

		public override string ToString()
		{
			return nameValue;
		}

		public static Direction valueOf(string name)
		{
			foreach (Direction enumInstance in Direction.valueList)
			{
				if (enumInstance.nameValue == name)
				{
					return enumInstance;
				}
			}
			throw new System.ArgumentException(name);
		}
	}

}

File: logisim/app/src/main/java/logisim/data/Direction.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.data;

public enum Direction implements AttributeOptionInterface {

	East(0), North(1), West(2), South(3);

	private final int id;

	Direction(int id) {
		this.id = id;
	}

	/**
	 * @deprecated Please use valueOf instead
	 */
	@Deprecated
	public static Direction parse(String str) {
		String cap = str.substring(0, 1).toUpperCase() + str.substring(1);
		return Direction.valueOf(cap);
	}

	public static final Direction[] cardinals = { East, North, West, South };

	public static Direction[] getCardinals() {
		return cardinals;
	}

	public static Direction fromInt(int x) {
		return cardinals[x];
	}

	private static final String[] displayStrings = { Strings.get("directionEastOption"), Strings.get("directionNorthOption"),
			Strings.get("directionWestOption"), Strings.get("directionSouthOption") };

	public String toDisplayString() {
		return displayStrings[id];
	}

	private static final String[] displayStringsVertical = { Strings.get("directionEastVertical"),
			Strings.get("directionNorthVertical"), Strings.get("directionWestVertical"),
			Strings.get("directionSouthVertical") };

	public String toVerticalDisplayString() {
		return displayStringsVertical[id];
	}

	public double toRadians() {
		return id * Math.PI / 2.0;
	}

	public int toDegrees() {
		return id * 90;
	}

	public Direction reverse() {
		return fromInt((id + 2) & 3);
	}

	public Direction rotateCW() {
		return fromInt((id - 1) & 3);
	}

	public Direction rotateCCW() {
		return fromInt((id + 1) & 3);
	}

	// for AttributeOptionInterface
	public Object getValue() {
		return this;
	}
}
